enterprise/server/scheduling/task_leaser/task_leaser.go

```diff
@@ -29,36 +32,74 @@ const (
 )
 
 type TaskLeaser struct {
-	env               environment.Env
-	executorID        string
-	taskID            string
-	leaseID           string
-	supportsReconnect bool
-	quit              chan struct{}
-	mu                sync.Mutex // protects stream
-	stream            scpb.Scheduler_LeaseTaskClient
-	ttl               time.Duration
-	cancelFunc        context.CancelFunc
+	env        environment.Env
+	executorID string
 }
 
-func NewTaskLeaser(env environment.Env, executorID string, taskID string) *TaskLeaser {
+var _ interfaces.TaskLeaser = (*TaskLeaser)(nil)
+
+func NewTaskLeaser(env environment.Env, executorID string) *TaskLeaser {
 	return &TaskLeaser{
 		env:        env,
 		executorID: executorID,
+	}
+}
+
+func (t *TaskLeaser) Lease(ctx context.Context, taskID string) (interfaces.TaskLease, error) {
+	lease := &TaskLease{
+		env:        t.env,
+		executorID: t.executorID,
 		taskID:     taskID,
+		execTask:   &repb.ExecutionTask{},
 		quit:       make(chan struct{}),
 		ttl:        100 * time.Second,
 	}
+	ctx, serializedTask, err := lease.claim(ctx)
+	if err != nil {
+		return nil, err
+	}
+	lease.ctx = ctx
+	if err := proto.Unmarshal(serializedTask, lease.execTask); err != nil {
+		lease.Close(ctx, nil, false /*=retry*/)
+		return nil, status.InternalErrorf("unmarshal ExecutionTask: %s", err)
+	}
+	return lease, nil
+}
+
+type TaskLease struct {
+	env        environment.Env
+	executorID string
+	taskID     string
+
+	ctx               context.Context
+	execTask          *repb.ExecutionTask
+	leaseID           string
+	supportsReconnect bool
+	quit              chan struct{}
+	mu                sync.Mutex // protects stream
+	stream            scpb.Scheduler_LeaseTaskClient
+	ttl               time.Duration
+	cancelFunc        context.CancelFunc
+}
+
+var _ interfaces.TaskLease = (*TaskLease)(nil)
```

Comment:
nit: move to top of the file and add a comment like:

// Verify that TaskLease implements the interfaces.TaskLease interface.
var _ interfaces.TaskLease = (*TaskLease)(nil)

Reply:
(nit: same with other check -- yes i am reviewing this cl backwards)

Reply:
Done